File: 19th/코딩테스트 심화반/week6/15900번_나무 탈출.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

vector<int> adj[500001]; // 트리를 인접리스트로 저장
bool visited[500001]; // 방문 여부

int main() {
    // 빠른 입출력
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int N;
    cin >> N;

    int n1, n2;
    for (int i = 0; i < N - 1; i++) {
        cin >> n1 >> n2;
        adj[n1].push_back(n2);
        adj[n2].push_back(n1);
    }

    queue<pair<int, int>> q;

    q.push({ 1, 0 }); // 루트 노드부터 시작
    visited[1] = true;

    int cnt = 0;
    while (!q.empty()) {
        int n = q.front().first;
        int dist = q.front().second;
        q.pop();

        bool leaf = true; 
        for (auto v : adj[n]) { // n번 노드와 연결된 노드 v
            if (!visited[v]) {
                leaf = false;
                visited[v] = true;
                q.push({ v, dist + 1 });
            }
        }
        if (leaf) cnt += dist; // 만약 리프 노드라면 루트로부터의 거리를 저장
    }

    if (cnt % 2 == 1) cout << "Yes";
    else cout << "No";
}

File: 19th/코딩테스트 심화반/week6/15900번_나무 탈출.py
from collections import deque

n = int(input())

adj = [[] for _ in range(n + 1)]  # 인접 리스트
visited = [0] * (n + 1)

for _ in range(n-1):
    a, b = map(int, input().split())
    adj[a].append(b)
    adj[b].append(a)

q = deque()
cnt = 0  # 리프 노드까지의 거리 합
visited[1] = 1  # 시작 노드 방문 표시
q.append((1, 0))  # (현재 노드, 거리)

while q:
    num, dist = q.popleft()

    leaf = True  # 현재 노드가 리프 노드인지 판별
    for v in adj[num]:
        if not visited[v]:
            leaf = False
            visited[v] = 1
            q.append((v, dist + 1))
    
    if leaf:  # 리프 노드라면
        cnt += dist  # 거리 누적

# 리프 노드까지의 거리 합(cnt)이 홀수면 이길 수 있음
if cnt % 2 == 1:
    print("Yes")
else:
    print("No")